File: 295/report/timsort_test.py
import random
import time
import numpy as np

# L1 - 16,384 = 2^14
# L3 - 1,048,576 = 2^20
# Overfit - 16,777,216 = 2^24 = 16777216
sizes = [16384, 1048576, 16777216]
res = ["","",""]

# warmups
for i in range(4):
    SIZE = 1000000
    print("warmup...")
    data = np.array([random.randint(0, SIZE) for _ in range(SIZE)], dtype=np.int64)
    data.sort()

# TEST RANDOM ARRAYS
TYPE = "RANDOM"
RUNCOUNT = 4
for cache, cache_size in enumerate(sizes):
    print(f"running tests on size {cache_size}")
    for run in range(RUNCOUNT):
        # use numpy type cast into 64-bit ints
        # probably not identical to C types but hopefully it is "closer"
        data = np.array([random.randint(0, cache_size) for _ in range(cache_size)], dtype=np.int64)
        # measure only the sort time
        start_time = time.time()
        data.sort() # timsort native
        end_time = time.time()
        run_time = end_time - start_time
        concat = "" if run+1==RUNCOUNT else ", "
        res[cache] += f"{run_time}{concat}"
        print(f"Run {run+1} cache {cache}, results: {run_time} seconds")

print("\n\nresults:\n")
[print(f"[{r}]") for r in res]
print(f"testing {TYPE} arrays complete!")

File: 295/report/timsort_test.c
// Enable clock_gettime: https://stackoverflow.com/a/40515669
#define _POSIX_C_SOURCE 199309L

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <time.h>
#include "helpers.c"
#include "timsort/timsort.h"

// Comparison function for int64_t
int compareInt64(const void* a, const void* b) {
    int64_t int_a = *(int64_t*)a;
    int64_t int_b = *(int64_t*)b;
    if (int_a < int_b) return -1;
    else if (int_a > int_b) return 1;
    else return 0;
}

void timSort(int64_t arr[], int n) {
    timsort(arr, n, sizeof(int64_t), compareInt64);
}

void warmup() {
    // Warmups
    for (int i = 0; i < 4; i++) {        
        printf("Warmup...\n");
        int64_t* data = random_array(SIZE);
        timSort(data, SIZE);
        free(data);
    }
}

void testRandomArrays(char *res) {
    printf("Testing timSort on random arrays\n");
    for (int i = 0; i < RUNCOUNT; i++) {
        int64_t* data = random_array(SIZE);
        
        char* descr = "random timsort";
        double run_time = 0;
        TIMING_RESULT(descr, timSort(data, SIZE), run_time);

        // Append the run time to the result string
        char buffer[50];
        snprintf(buffer, sizeof(buffer), "%f%s", run_time, (i + 1 == RUNCOUNT) ? "" : ", ");
        strcat(res, buffer);
        
        // printf("Run %d results: %f ms\n", i + 1, run_time);
        free(data);
    }
    printf("Testing timsort sort on random arrays complete!\n");
}

void testSortedArrays(char *res) {
    printf("Testing timSort on sorted arrays\n");
    for (int i = 0; i < RUNCOUNT; i++) {
        int64_t* data = sorted_array(SIZE);
        
        char* descr = "sorted timsort";
        double run_time = 0;
        TIMING_RESULT(descr, timSort(data, SIZE), run_time);

        // Append the run time to the result string
        char buffer[50];
        snprintf(buffer, sizeof(buffer), "%f%s", run_time, (i + 1 == RUNCOUNT) ? "" : ", ");
        strcat(res, buffer);
        
        // printf("Run %d results: %f ms\n", i + 1, run_time);
        free(data);
    }
    printf("Testing timsort sort on sorted arrays complete!\n");
    // printf("\nRandom array results:\n[%s]\n", rand_results);
}

void testRevSortedArrays(char *res) {
    printf("Testing timSort on reverse sorted arrays\n");
    for (int i = 0; i < RUNCOUNT; i++) {
        int64_t* data = revsorted_array(SIZE);
        
        char* descr = "reverse sorted timsort";
        double run_time = 0;
        TIMING_RESULT(descr, timSort(data, SIZE), run_time);

        // Append the run time to the result string
        char buffer[50];
        snprintf(buffer, sizeof(buffer), "%f%s", run_time, (i + 1 == RUNCOUNT) ? "" : ", ");
        strcat(res, buffer);
        free(data);
    }
    printf("Testing timsort sort on reverse sorted arrays complete!\n");
}

int main() {
    srand(time(NULL));
    printf("\nTesting timsort Sort:\n");
    warmup();    

    // test
    char revprint[2048] = "";
    testRevSortedArrays(revprint);
    char sortprint[2048] = "";
    testSortedArrays(sortprint);
    char randprint[2048] = "";
    testRandomArrays(randprint);
    
    printf("\nSorted array results:\n[%s]\n", sortprint);
    printf("\nRev Sort array results:\n[%s]\n", revprint);   
    printf("\nRandom array results:\n[%s]\n", randprint);  
    return 0;
}